File: haskell/euler23.hs
--These solutions to project euler challenges by no means illustrate graceful or robust code, but are mostly hack-jobs intent on simply reaching the answer!!

--divisors :: Integer -> [Integer]

--divisors 1 = [1]

--divisors x = 1:[ y | y <- [2..(x `div` 2)], x `mod` y == 0] 


minus :: (Eq a) => [a] -> [a] -> [a]
minus [] xs                      = []
minus (y:ys) xs | y `notElem` xs = y : (minus ys xs)
                | otherwise      = minus ys xs

isDivisor :: Int -> Int -> Bool
isDivisor x y 
   | x `mod` y == 0 && x /= y = True
   | otherwise                = False

sumOfDivisors :: Int -> Int
sumOfDivisors x = sum $ filter (isDivisor x) [1..x]

getAbundantNumbers :: [Int]
getAbundantNumbers = filter (\y -> y < sumOfDivisors y) [1..28123]

sumList = [ x+y | x <- getAbundantNumbers, y<-getAbundantNumbers, x + y <= 28123]

excludedSumList = [1..28123] `minus` sumList 

main = print (sum excludedSumList)

File: ruby/euler23.rb
#These solutions to project euler challenges by no means illustrate graceful or robust code, but are mostly hack-jobs intent on simply reaching the answer!!

#require "prime.rb"


class Integer

	def factors() (1..self-1).select { |n| (self % n).zero? } end

	def sumfact() self.factors.reduce(:+) end

	def is_abundant()  self.sumfact > self end

end


abundants =  (2..20161).select {|x| x.is_abundant}



i = 0
sums = []
abundants.each do |x|
	abundants[i..abundants.length].each do |y|
		sum = x + y
		sums << sum unless sum > 20161
	end
	i += 1
end
sums.uniq!

#p (1..28123).reject {|x| sums.include? x}.reduce(:+)
p ([*1..20161] - sums).reduce(:+)




